cluster-autoscaler/proposals/pricing.md

```diff
@@ -0,0 +1,177 @@
+# Cost-based node group ranking function for Cluster Autoscaler
+##### Author: mwielgus
+ 
+# Introduction
+ 
+Cluster autoscaler tries to increase the cluster size if there are some pods that are unable to fit into the nodes currently present in the cluster. If there are two or more types of nodes in the cluster CA has to decide which one to add. At this moment it, by default, just picks a random one (there are some other expanders but they are also relatively simple).
+So it may add a 32 cpu node with gpu and 500 gb of ram to accommodate a pod that requires just a single node with a bit of memory. In order to properly support heterogeneous clusters we need to properly choose a node group and decide which one to expand, and be able to tell expensive expansion option from the cost effective one.
+ 
+# Node cost estimation
+ 
+To correctly choose the cheapest option from the set of available options we need to be able to calculate the cost of a single node. For example, for GKE/GCE node the cost is well known, however these numbers are not available through the api, so some short config file would be probably needed. As we don’t do real billing here but try to estimate the differences between the cluster expansion options the numbers don’t have to be super exact. CA should just price the cheaper instances lower than the more expensive ones.
+ 
+# Choosing the best node pool to expand
+ 
+Knowing the cost of a single node is only a part of the story. We need to choose the best node pool to accommodate the unscheduled pods, as a whole. However putting all of the pending pods to a single node pool may not be always an option because:
+ 
+* Node pool min/max boundaries. Pending pods may require expanding it beyond the max range.
+* Some pods may not fit into the machines.
+* Different node selectors.
+ 
+Different node pools may accept different pods and accommodate them on different number of nodes what can result in completely different cost. Let's denote the costs of the expansion as C.
+
+Let me give you an example.
+ 
+* Option1:  requires 3 nodes of Type1, accommodates pods P1, P2, P3 and costs C1=10$
+* Option2:  requires 2 nodes of Type2, accommodates pods P1, P3, P4, P5 and costs C2=20$. 
+ 
+It is hard to tell whether we will get a better deal with paying 10$ for having 3 pods running or with paying 20$ for a different set of pods. We need to make C1 and C2 somehow comparable.
+ 
+We can compute how much would it cost to run a pod on a machine that is perfectly suited to its needs. From GKE pricing we know that:
```

Comment:
Makes sense, just a note for the future: I don't think AWS has a very good API for fetching the current bid for spot instances. There's an API action (`GetEC2SpotPriceHistory`) to get the bidding price over some period of time, but it's not necessarily the price you end up paying. I'm sure we can get some average over time to work around it, bringing it up as it can be problematic in the future. 

@mumoshu may know more on this topic.